### PR TITLE
Add support for FilterMaintenanceAccessory

### PR DESCRIPTION
**Pull request by gjvanderheiden**
[#124](https://github.com/hap-java/HAP-Java/pull/124)

# Pull Request Checklist

Please confirm that you've done the following when opening a new pull request:

  - [x] For fixes and other improvements, please reference the GitHub issue that your change addresses.
  - [x] For fixes, optimizations and new features, please add an entry to the CHANGES.md file.
  - [x] Run mvn compile before committing, so that the auto-code formatter will format your changes consistently with the rest of the project.

### CODE REVIEW

**Comment by gjvanderheiden** [comment]
linked issue : [#123](https://github.com/hap-java/HAP-Java/issues/123)

Had to change the IntegerCharacteristic too. It didn't support unit nor no value. (needed for ResetFilter). Tested it with my own project, which uses this characteristic excessively.

Really missing exciting unit tests here. Unit test could test how the IntegerCharacteristic is responding to this in the json build.

File: src/main/java/io/github/hapjava/accessories/AirPurifierAccessory.java [hunk 1]
```diff
@@ -79,8 +79,23 @@
   /** Unsubscribes from changes in the target state of the air purifier. */
   void unsubscribeTargetState();
 
+  /**
+   * If a filter maintenance service is needed as a linked service to this AirPurifier, this is the
+   * place.
+   *
+   * @return an instance of FilterMaintenanceAccessory, null if not needed.
+   */
+  default FilterMaintenanceAccessory getFilterMaintenanceAccessory() {
```
Comment (yfre) [inline src/main/java/io/github/hapjava/accessories/AirPurifierAccessory.java:88]: why we need it here?
the filter service can be linked using linked Service approach. i.e. not additional method need it for it

File: src/main/java/io/github/hapjava/accessories/AirPurifierAccessory.java [hunk 2]
```diff
@@ -79,8 +79,23 @@
...
+    return null;
+  };
+
   @Override
   default Collection<Service> getServices() {
-    return Collections.singleton(new AirPurifierService(this));
+    AirPurifierService service = new AirPurifierService(this);
```
Comment (yfre) [inline src/main/java/io/github/hapjava/accessories/AirPurifierAccessory.java:94]: linked service approach is generic one, we dont need to change getService here. 
we can link any service with any other service, e.g. we can add filter service but also Fan or Slat

File: src/main/java/io/github/hapjava/characteristics/impl/base/IntegerCharacteristic.java [hunk 3]
```diff
@@ -62,17 +62,21 @@ public IntegerCharacteristic(
...
+                builder.add("unit", unit);
+              }
+              return builder;
             });
   }
 
   @Override
   protected CompletableFuture<Integer> getValue() {
-    return getter.map(integerGetter -> integerGetter.get()).get();
+    if (getter.isPresent()) {
```
Comment (yfre) [inline src/main/java/io/github/hapjava/characteristics/impl/base/IntegerCharacteristic.java:75]: good point with null, 
replacing get with orElse should do the trick
`return getter.map(integerGetter -> integerGetter.get()).orElse(null);`

**Review by yfre — CHANGES_REQUESTED** [review]
for filter maintenance we need additional characteristics but i could not find them in PR.
e.g.
ResetFilterIndicationCharacteristic
FilterLifeLevelCharacteristic
FilterChangeIndicationCharacteristic

**Comment by yfre** [comment]
with the new linked service support we dont need to have any dedicated methods on accessory levels but can basically add any service to any accessory, e.g. we can add Fan to Air Purifier
it looks like this (im adding battery service to switch but it works the same for all other accessories and services)

```
final MockSwitch switch1 = new MockSwitch();
final MockBatteryService batteryService = new MockBatteryService();
switch1.getServices().add(batteryService.getPrimaryService());
```

means the only thing we need to implement is the FilterMaintanceService

File: src/main/java/io/github/hapjava/characteristics/impl/base/IntegerCharacteristic.java (hunk 3 above)
Comment (gjvanderheiden) [reply src/main/java/io/github/hapjava/characteristics/impl/base/IntegerCharacteristic.java:75]: Thank you, I was struggling with that one

File: src/main/java/io/github/hapjava/accessories/AirPurifierAccessory.java (hunk 2 above)
Comment (gjvanderheiden) [reply src/main/java/io/github/hapjava/accessories/AirPurifierAccessory.java:94]: I doubted on this one while adding it. But, I preserved the declarational properties of this lib. I mean, everything is done with templating / interfaces. Like the new feature #115(https://github.com/hap-java/HAP-Java/pull/115) this very generic characteristic is only possible when HAP-java allows it. That's not necessarily a bad thing, but a choice. It protects the user of HAP-java from adding the wrong stuff. It show's the limits of HAP in the interfaces. 

Leaving this out allong with getFilterMaintenanceAccessory() mean that you'll have to override getServices if you wan't to use it (not done by hab-java). That is the only alternative I could see.

What do you think?

File: src/main/java/io/github/hapjava/accessories/AirPurifierAccessory.java (hunk 1 above)
Comment (gjvanderheiden) [reply src/main/java/io/github/hapjava/accessories/AirPurifierAccessory.java:88]: is the same issue as the next one as I see it.

**Comment by gjvanderheiden** [comment]
> for filter maintenance we need additional characteristics but i could not find them in PR.
> e.g.
> ResetFilterIndicationCharacteristic
> FilterLifeLevelCharacteristic
> FilterChangeIndicationCharacteristic

buggar. beep**-ed something up with the change lists in my IDE. Thanks for mentioning

**Comment by gjvanderheiden** [comment]
Now ResetFilterIndicationCharacteristic is kaput... will fix that first.

> with the new linked service support we dont need to have any dedicated methods on accessory levels but can basically add any service to any accessory, e.g. we can add Fan to Air Purifier
> it looks like this (im adding battery service to switch but it works the same for all other accessories and services)
> 
> ```
> final MockSwitch switch1 = new MockSwitch();
> final MockBatteryService batteryService = new MockBatteryService();
> switch1.getServices().add(batteryService.getPrimaryService());
> ```
> 
> means the only thing we need to implement is the FilterMaintanceService
I like the dynamic possibilities.
I could be missing something, but getServices() returns a new instance of the service with every call. I can't see this working with HomekitBridge.addAccessory().

**Comment by gjvanderheiden** [comment]
Another alternative would be:

HomekitAccessory accessory = new MockSwitch(...);
accessory.addLinkedAccessory(new MockBatteryAccessory());

**Comment by gjvanderheiden** [comment]
@yfre thrown the 2 modification out of the window.  How to link the services is another issue. Thank you for the review

**Comment by gjvanderheiden** [comment]
added comment to #104(https://github.com/hap-java/HAP-Java/issues/104). Can't add the accessory right now.

**Review by yfre — APPROVED** [review]
looks good for me.

**Pull request merged**